source/common/stats/thread_local_store.cc

```diff
@@ -156,6 +156,26 @@ void ThreadLocalStoreImpl::clearScopeFromCaches(uint64_t scope_id) {
   }
 }
 
+absl::string_view ThreadLocalStoreImpl::truncateStatNameIfNeeded(absl::string_view name) {
+  // If the main allocator requires stat name truncation, warn and truncate, before
+  // attempting to allocate.
+  if (alloc_.requiresBoundedStatNameSize()) {
+    const uint64_t max_length = stats_options_.maxNameLength();
+
+    // Note that the heap-allocator does not truncate itself; we have to
+    // truncate here if we are using heap-allocation as a fallback due to an
+    // exahusted shared-memroy block
+    if (name.size() > max_length) {
+      ENVOY_LOG_MISC(
+          warn,
+          "Statistic '{}' is too long with {} characters, it will be truncated to {} characters",
+          name, name.size(), max_length);
+      name = absl::string_view(name.data(), max_length);
+    }
+  }
+  return name;
```

Comment:
Random thought that is unrelated to this PR (existing issue). How confident are we that two truncated stats won't conflict, e.g. if I truncated `a.b.c.d` and `a.b.c.e` to 5 characters, we get  overlap.. it seems this is double bad, since we're both truncating and conflating..

Reply:
Yeah...in the past if I wanted to truncate to 'n' I'd actually truncate to n-k, and use a k-character base64 hash of the untruncated string to disambiguate.  That would still be pretty awful for stats as you'd have no idea which one you were looking at. Basically, truncation is bad and should be avoided.